packages/runtime/datastore/src/dataStoreRuntime.ts

```diff
@@ -650,66 +652,166 @@ export class FluidDataStoreRuntime
 		);
 	}
 
-	public process(
-		message: ISequencedDocumentMessage,
+	/**
+	 * Process channel messages. The messages here are contiguous channel types messages in a batch for this data
+	 * store.
+	 * @param message - The core message with common properties for all the messages.
+	 * @param messageContents - The contents of the channel messages and their metadata.
+	 * @param local - true if the messages were originally generated by the client receiving it.
+	 */
+	private processChannelMessages(
+		message: ISequencedRuntimeMessageCore,
+		messageContents: IRuntimeMessageContents[],
 		local: boolean,
-		localOpMetadata: unknown,
 	) {
 		this.verifyNotClosed();
 
+		/*
+		 * Bunch contiguous messages for the same channel and send them together.
+		 * This is an optimization where DDSes can process a bunch of ops together. DDSes
+		 * like merge tree or shared tree can process ops more efficiently when they are bunched together.
+		 */
+		let currentAddress: string | undefined;
+		let currentMessageContents: IRuntimeMessageContents[] = [];
+
+		const sendBunchedMessages = () => {
+			assert(currentAddress !== undefined, "an address must exist to send messages");
+
+			// process the last set of channel ops
+			const channelContext = this.contexts.get(currentAddress);
+			assert(!!channelContext, "Channel context not found");
+
+			// For now, send the message to be processed one by one. This will be updated to send the bunch later.
+			for (const {
+				contents,
+				clientSequenceNumber,
+				localOpMetadata,
+			} of currentMessageContents) {
+				channelContext.processOp(
+					{ ...message, contents, clientSequenceNumber },
+					local,
+					localOpMetadata,
+				);
+			}
+			currentMessageContents = [];
+		};
+
+		for (const { contents, ...restOfMessageContents } of messageContents) {
+			const envelope = contents as IEnvelope;
+
+			// If the address of the message changes while processing the batch, send the current bunch.
+			if (currentAddress !== undefined && currentAddress !== envelope.address) {
```

Comment:
I think currentAddress  === undefined is not interesting case (there are no messages), and it can't become undefined (we should fail sooner). I'd remove this check for easier readability

Reply:
That check is required when the first message is processed because we don't want to send it. I moved it inside `sendBunchedMessages` though.